akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala

```diff
@@ -442,6 +442,29 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
         s"toSequenceNr [$toSequenceNr] must be less than or equal to lastSequenceNr [$lastSequenceNr]"), toSequenceNr)
   }
 
+  /**
+   * INTERNAL API.
+   * An [[Eventsourced]] actor can request cleanup by deleting either a range of, or all persistent events.
+   * For example, on successful snapshot completion, delete messages within a configurable `snapshotAfter`
+   * range that are less than or equal to the given [[SnapshotMetadata.sequenceNr]]
+   * (provided the [[SnapshotMetadata.sequenceNr]] is <= to [[Eventsourced#lastSequenceNr]]).
+   *
+   * Or delete all by using `Long.MaxValue` as the `toSequenceNr`
+   * {{{ m.copy(sequenceNr = Long.MaxValue) }}}
+   */
+  private[akka] def deleteMessages(
+    e:               SaveSnapshotSuccess,
+    keepNrOfBatches: Int,
+    snapshotAfter:   Int): Unit = {
+    /* Delete old events but keep the latest around
+      1. It's not safe to delete all events immediately because snapshots are typically stored with
+         a weaker consistency level. A replay might "see" the deleted events before it sees the stored
+         snapshot, i.e. it could use an older snapshot and not replay the full sequence of events
+      2. If there is a production failure, it's useful to be able to inspect the events while debugging */
+    val sequenceId = e.metadata.sequenceNr - keepNrOfBatches * snapshotAfter
+    if (sequenceId > 0) deleteMessages(sequenceId)
+  }
```

Comment:
Ensure there is still a `SaveSnapshotSuccess` requirement within the context of calling 

```private[akka] def deleteMessages``` 

to retain the notion of 

```scala
def receiveSnapshotResult: Receive = {
  case SaveSnapshotSuccess(meta) => deleteMessages...
}

def receiveCommand: Receive = {
  case SaveSnapshotSuccess(meta) => deleteMessages...
}
```

Reply:
I think this is one of those tricky places where we cannot add methods, because user actors will inherit it and may have a method called `deleteMessages` already, I think you have to make it a shared method in sharding rather.

Reply:
Works for me :) Updating..

Reply:
In general though, the snapshot and messages logic around this and related is too scattered and I think should be consolidated.

Reply:
although unlikely this is not fully **source** compatible because it could clash if user has defined a method already with exactly the same signature

by naming it `internalDeleteMessagesBeforeSnapshot` we reduce the risk enough to be ok

Reply:
I find it interesting that:

`PersistentShard`:
```scala
case e: SaveSnapshotSuccess =>
    val sequenceId = e.metadata.sequenceNr - keepNrOfBatches * snapshotAfter
    if (sequenceId > 0) deleteMessages(sequenceId)
  
```

`PersistentShardCoordinator`:
```scala
 case e: DeleteMessagesSuccess =>
    val deleteTo = e.toSequenceNr - 1
    val deleteFrom = math.max(0, deleteTo - (keepNrOfBatches * snapshotAfter))
    deleteSnapshots(SnapshotSelectionCriteria(
      minSequenceNr = deleteFrom,
      maxSequenceNr = deleteTo
    ))
```

`RemoveInternalClusterShardingData`'s `RemoveOnePersistenceId`:
```scala
case RecoveryCompleted =>
  deleteMessages(Long.MaxValue)
  if (hasSnapshots) deleteSnapshots(SnapshotSelectionCriteria())
```

Just strikes me as scattered ?

Reply:
`RemoveInternalClusterShardingData` is a self contained main program with the purpose to remove all events and all snapshots. I don't think that one has any similarities with PersistentShard and PersistentShardCoordinator

Reply:
Thanks